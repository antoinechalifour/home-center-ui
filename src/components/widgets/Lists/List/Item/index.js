import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql, compose } from 'react-apollo'
import { updateListItem, deleteListItem } from 'queries/lists'
import Item from './Item'

class ItemContainer extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
    updateListItem: PropTypes.func.isRequired,
    deleteListItem: PropTypes.func.isRequired
  }

  _onToggleStatus = () => {
    this.props.updateListItem({
      variables: {
        input: {
          id: this.props.id,
          text: this.props.text,
          done: !this.props.done
        }
      }
    })
  }

  _onUpdateText = value => {
    this.props.updateListItem({
      variables: {
        input: {
          id: this.props.id,
          text: value,
          done: this.props.done
        }
      }
    })
  }

  _onDelete = () => {
    this.props.deleteListItem({
      variables: {
        input: { id: this.props.id }
      }
    })
  }

  render () {
    return (
      <Item
        {...this.props}
        updateText={this._onUpdateText}
        toggleStatus={this._onToggleStatus}
        deleteItem={this._onDelete}
      />
    )
  }
}

const updateListItemOptions = {
  name: 'updateListItem'
}

const deleteListItemOptions = {
  name: 'deleteListItem'
}

export default compose(
  graphql(updateListItem, updateListItemOptions),
  graphql(deleteListItem, deleteListItemOptions)
)(ItemContainer)
