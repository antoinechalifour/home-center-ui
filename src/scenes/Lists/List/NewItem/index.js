import { graphql } from 'react-apollo'
import { addListItem } from 'queries/lists'
import NewItem from './NewItem'

const queryOptions = {}

export default graphql(addListItem, queryOptions)(NewItem)
