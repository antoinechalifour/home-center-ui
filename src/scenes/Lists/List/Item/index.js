import { graphql, compose } from 'react-apollo'
import { updateListItem, deleteListItem } from 'queries/lists'
import Item from './Item'

const updateListItemOptions = {
  name: 'updateListItem',
  options: {
    refetchQueries: ['GetList']
  }
}

const deleteListItemOptions = {
  name: 'deleteListItem',
  options: {
    refetchQueries: ['GetList']
  }
}

export default compose(
  graphql(updateListItem, updateListItemOptions),
  graphql(deleteListItem, deleteListItemOptions)
)(Item)
