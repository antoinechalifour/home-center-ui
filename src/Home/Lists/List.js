import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { graphql, compose } from 'react-apollo'
import gql from 'graphql-tag'
import DeleteIcon from 'react-icons/lib/md/clear'
import Card from 'ui/Card'
import Item from './Item'
import NewItem from './NewItem'

const Wrapper = styled(Card)`
  & + & {
    margin-top: 12px;
  }
`

const Name = styled.div`
  text-transform: uppercase;
  padding: 12px;
  text-align: center;
  box-shadow: 0 1px 3px rgba(0, 0, 0, .13);
  position: relative;

  svg {
    opacity: .54;
    cursor: pointer;
    position: absolute;
    right: 12px;
    top: 50%;
    transform: translateY(-50%);
  }
`

const List = ({ id, name, data, mutate }) => {
  const isLoading = data.loading

  if (isLoading) {
    return null
  }

  return (
    <Wrapper>
      <Name>
        <span>{name}</span>
        <DeleteIcon onClick={() => mutate({ variables: { id } })} />
      </Name>
      <ul>
        {data.list.items.map(x => <Item key={x.id} {...x} />)}
      </ul>
      <NewItem listId={id} />
    </Wrapper>
  )
}

List.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    list: PropTypes.shape({
      items: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          text: PropTypes.string.isRequired,
          done: PropTypes.bool.isRequired
        })
      ).isRequired
    })
  }).isRequired,
  mutate: PropTypes.func.isRequired
}

export default compose(
  graphql(
    gql`
      query ListQuery ($id: Int) {
        list (id: $id) {
          items {
            id,
            text,
            done
          }
        }
      }
  `,
    {
      options: ({ id }) => ({
        variables: { id }
      })
    }
  ),
  graphql(
    gql`
      mutation deleteList($id: Int) {
        deleteList(id: $id) {
          id
        }
      }
    `,
    {
      options: {
        refetchQueries: ['ListsQuery']
      }
    }
  )
)(List)
