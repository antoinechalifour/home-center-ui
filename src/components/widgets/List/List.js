import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import DeleteIcon from 'react-icons/lib/md/clear'
import Editable from 'components/Editable'
import gqlLoaderHoc from 'components/GqlLoader'
import Item from './Item'
import NewItem from './NewItem'

export function List ({ id, data, updateName, deleteList }) {
  return (
    <Container>
      <Name>
        <Editable onChange={updateName}>
          {data.list.name}
        </Editable>
        <DeleteIcon onClick={deleteList} />
      </Name>
      <ul>
        {data.list.items.map(x => <Item key={x.id} {...x} />)}
      </ul>
      <NewItem listId={id} />
    </Container>
  )
}

List.propTypes = {
  id: PropTypes.number.isRequired,
  data: PropTypes.shape({
    list: PropTypes.shape({
      name: PropTypes.string.isRequired,
      items: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          text: PropTypes.string.isRequired,
          done: PropTypes.bool.isRequired
        })
      ).isRequired
    })
  }).isRequired,
  updateName: PropTypes.func.isRequired,
  deleteList: PropTypes.func.isRequired
}

export default gqlLoaderHoc(List)

const Container = styled.div`
`

const Name = styled.div`
  text-transform: uppercase;
  padding: 8px 12px;
  position: relative;
  background: ${({ theme }) => theme.colors.primary};
  color: #fff;

  svg {
    cursor: pointer;
    position: absolute;
    right: 12px;
    top: 50%;
    transform: translateY(-50%);
  }
`
