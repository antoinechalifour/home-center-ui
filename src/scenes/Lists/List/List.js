import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import DeleteIcon from 'react-icons/lib/md/clear'
import Editable from 'components/Editable'
import Card from 'ui/Card'
import Item from './Item'
import NewItem from './NewItem'

export default function List ({ id, data, updateList, deleteList }) {
  const isLoading = data.loading

  if (isLoading) {
    return null
  }

  return (
    <Wrapper>
      <Name>
        <Editable
          onChange={value =>
            updateList({
              variables: {
                input: { id, name: value }
              }
            })}
        >
          {data.list.name}
        </Editable>
        <DeleteIcon
          onClick={() =>
            deleteList({
              variables: {
                input: { id }
              }
            })}
        />
      </Name>
      <ul>
        {data.list.items.map(x => <Item key={x.id} {...x} />)}
      </ul>
      <NewItem listId={id} />
    </Wrapper>
  )
}

List.propTypes = {
  id: PropTypes.number.isRequired,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    list: PropTypes.shape({
      name: PropTypes.string.isRequired,
      items: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          text: PropTypes.string.isRequired,
          done: PropTypes.bool.isRequired
        })
      ).isRequired
    })
  }).isRequired,
  updateList: PropTypes.func.isRequired,
  deleteList: PropTypes.func.isRequired
}

const Wrapper = styled(Card)`
`

const Name = styled.div`
  text-transform: uppercase;
  padding: 12px;
  text-align: center;
  border-bottom: 1px solid rgba(0, 0, 0, .15);
  position: relative;

  svg {
    opacity: .24;
    cursor: pointer;
    position: absolute;
    right: 12px;
    top: 50%;
    transform: translateY(-50%);
  }
`
