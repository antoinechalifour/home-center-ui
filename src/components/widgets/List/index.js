import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql, compose } from 'react-apollo'
import {
  getList,
  updateList,
  deleteList,
  listUpdated,
  listItemCreated,
  listItemUpdated,
  listItemDeleted
} from 'queries/lists'
import List from './List'

export class ListContainer extends Component {
  static propTypes = {
    listId: PropTypes.number.isRequired,
    data: PropTypes.shape({
      subscribeToMore: PropTypes.func.isRequired
    }).isRequired,
    updateList: PropTypes.func.isRequired,
    deleteList: PropTypes.func.isRequired
  }

  componentDidMount () {
    this.props.data.subscribeToMore({
      document: listUpdated,
      variables: {
        id: this.props.listId
      },
      updateQuery: this._onListUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemCreated,
      variables: {
        listId: this.props.listId
      },
      updateQuery: this._onListItemCreated
    })

    this.props.data.subscribeToMore({
      document: listItemUpdated,
      variables: {
        listId: this.props.listId
      },
      updateQuery: this._onListItemUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemDeleted,
      variables: {
        listId: this.props.listId
      },
      updateQuery: this._onListItemDeleted
    })
  }

  _onListUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updates = subscriptionData.data.listUpdated

    return {
      ...prev,
      list: {
        ...list,
        ...updates
      }
    }
  }

  _onListItemCreated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const newItem = subscriptionData.data.listItemCreated

    return {
      ...prev,
      list: {
        ...list,
        items: [...list.items, newItem]
      }
    }
  }

  _onListItemUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updatedItem = subscriptionData.data.listItemUpdated

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.map(x => (x.id === updatedItem.id ? updatedItem : x))
      }
    }
  }

  _onListItemDeleted (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const deletedItem = subscriptionData.data.listItemDeleted

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.filter(x => x.id !== deletedItem.id)
      }
    }
  }

  _onUpdateName = name => {
    this.props.updateList({
      variables: {
        input: {
          id: this.props.listId,
          name
        }
      }
    })
  }

  _onDeleteList = () => {
    this.props.deleteList({
      variables: {
        input: { id: this.props.listId }
      }
    })
  }

  render () {
    return (
      <List
        {...this.props}
        updateName={this._onUpdateName}
        deleteList={this._onDeleteList}
      />
    )
  }
}

const getListOptions = {
  options: ({ listId }) => ({
    variables: { id: listId }
  })
}

const updateListOptions = {
  name: 'updateList'
}

const deleteListOptions = {
  name: 'deleteList',
  options: {
    refetchQueries: ['GetWidgets']
  }
}

export default compose(
  graphql(getList, getListOptions),
  graphql(updateList, updateListOptions),
  graphql(deleteList, deleteListOptions)
)(ListContainer)
