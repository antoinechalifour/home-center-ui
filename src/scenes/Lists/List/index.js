import React, { Component } from 'react'
import { graphql, compose } from 'react-apollo'
import {
  getList,
  updateList,
  deleteList,
  listUpdated,
  listItemCreated,
  listItemUpdated,
  listItemDeleted
} from 'queries/lists'
import List from './List'

export class ListContainer extends Component {
  componentDidMount () {
    this.props.data.subscribeToMore({
      document: listUpdated,
      variables: {
        id: this.props.id
      },
      updateQuery: this.onListUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemCreated,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemCreated
    })

    this.props.data.subscribeToMore({
      document: listItemUpdated,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemDeleted,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemDeleted
    })
  }

  onListUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updates = subscriptionData.data.listUpdated

    return {
      ...prev,
      list: {
        ...list,
        ...updates
      }
    }
  }

  onListItemCreated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const newItem = subscriptionData.data.listItemCreated

    return {
      ...prev,
      list: {
        ...list,
        items: [...list.items, newItem]
      }
    }
  }

  onListItemUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updatedItem = subscriptionData.data.listItemUpdated

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.map(x => (x.id === updatedItem.id ? updatedItem : x))
      }
    }
  }

  onListItemDeleted (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const deletedItem = subscriptionData.data.listItemDeleted

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.filter(x => x.id !== deletedItem.id)
      }
    }
  }

  render () {
    return <List {...this.props} />
  }
}

const getListOptions = {
  options: ({ id }) => ({
    variables: { id }
  })
}

const updateListOptions = {
  name: 'updateList'
}

const deleteListOptions = {
  name: 'deleteList'
}

export default compose(
  graphql(getList, getListOptions),
  graphql(updateList, updateListOptions),
  graphql(deleteList, deleteListOptions)
)(ListContainer)
