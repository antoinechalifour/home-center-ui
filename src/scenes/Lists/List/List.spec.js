import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import List from './List'

const renderer = new ShallowRenderer()

describe('scenes/Lists/List', () => {
  it('Should render null when the data is loading', () => {
    const props = {
      id: 4,
      name: 'My todo list',
      data: {
        loading: true,
        list: {
          items: []
        }
      },
      updateList: jest.fn(),
      deleteList: jest.fn()
    }
    const tree = renderer.render(<List {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should render correctly', () => {
    const props = {
      id: 4,
      name: 'My todo list',
      data: {
        loading: false,
        list: {
          items: [
            {
              id: 1,
              text: 'Item #1',
              done: false
            },
            {
              id: 2,
              text: 'Item #2',
              done: true
            },
            {
              id: 3,
              text: 'Item #3',
              done: false
            }
          ]
        }
      },
      updateList: jest.fn(),
      deleteList: jest.fn()
    }
    const tree = renderer.render(<List {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should call the mutation when the delete icon is clicked', () => {
    const props = {
      id: 6,
      name: 'My todo list #2',
      data: {
        loading: false,
        list: {
          items: []
        }
      },
      updateList: jest.fn(),
      deleteList: jest.fn()
    }
    const wrapper = shallow(<List {...props} />)

    wrapper.find('MdClear').simulate('click')

    expect(props.deleteList.mock.calls.length).toEqual(1)
    expect(props.deleteList.mock.calls[0][0]).toEqual({
      variables: { id: 6 }
    })
  })
})
