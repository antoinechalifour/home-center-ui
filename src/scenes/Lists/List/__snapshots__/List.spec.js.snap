// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes/Lists/List Should render correctly 1`] = `
<Styled(styled.div)>
  <styled.div>
    <span>
      My todo list
    </span>
    <MdClear
      onClick={[Function]}
    />
  </styled.div>
  <ul>
    <Apollo(Apollo(Item))
      done={false}
      id={1}
      text="Item #1"
    />
    <Apollo(Apollo(Item))
      done={true}
      id={2}
      text="Item #2"
    />
    <Apollo(Apollo(Item))
      done={false}
      id={3}
      text="Item #3"
    />
  </ul>
  <Apollo(NewItem)
    listId={4}
  />
</Styled(styled.div)>
`;

exports[`scenes/Lists/List Should render null when the data is loading 1`] = `null`;
